enterprise/server/remote_execution/containers/ociruntime/ociruntime.go

```diff
@@ -567,28 +590,58 @@ func (c *ociContainer) createRootfs(ctx context.Context) error {
 		if len(children) == 0 {
 			continue
 		}
-		lowerDirs = append(lowerDirs, path)
-	}
-	// Create workdir and upperdir.
-	workdir := filepath.Join(c.bundlePath(), "tmp", "rootfs.work")
-	if err := os.MkdirAll(workdir, 0755); err != nil {
-		return fmt.Errorf("create overlay workdir: %w", err)
+		newLowerDirs := append(lowerDirs, path)
+		mergedWorkdir := filepath.Join(c.bundlePath(), "tmp", fmt.Sprintf("merged%d.work", len(c.mergedMounts)))
+		mergedUpperdir := filepath.Join(c.bundlePath(), "tmp", fmt.Sprintf("merged%d.upper", len(c.mergedMounts)))
+		mntOptsLen := tplLen + len(strings.Join(append(c.mergedMounts, newLowerDirs...), ":")) + max(
+			// mergedWorkdir and mergedUpperdir are always longer than workDir and upperdir.
+			// So this `max` is	not strictly necessary, but it's here to fend	off future changes.
+			len(mergedWorkdir)+len(mergedUpperdir),
+			len(workdir)+len(upperdir),
+		)
+		if len(newLowerDirs) == 1 || mntOptsLen <= maxMntOptsLength {
+			lowerDirs = newLowerDirs
+			continue
+		}
+
+		// If the total length of the lowerDirs exceeds the kernel page size,
+		// create a merged overlay mount to reduce the number of layers.
+		if err := os.MkdirAll(mergedWorkdir, 0755); err != nil {
+			return fmt.Errorf("create overlay workdir: %w", err)
+		}
+		if err := os.MkdirAll(mergedUpperdir, 0755); err != nil {
+			return fmt.Errorf("create overlay upperdir: %w", err)
+		}
+		merged := filepath.Join(c.bundlePath(), "tmp", fmt.Sprintf("merged%d", len(c.mergedMounts)))
+		if err := os.MkdirAll(merged, 0755); err != nil {
+			return fmt.Errorf("create overlay merged: %w", err)
+		}
+		slices.Reverse(lowerDirs)
+		mntOpts := fmt.Sprintf(optionsTpl, strings.Join(lowerDirs, ":"), mergedUpperdir, mergedWorkdir)
+		log.CtxDebugf(ctx, "Mounting merged overlayfs to %q, options=%q, len=%d", merged, mntOpts, len(mntOpts))
```

Comment:
nit: mind adding a sanity check here that the mount options length is <= 4095, and returning an error? In case we have some logic issue above (in the future), or in case we have a single mount dir that is really long, it would be nice to return a clear error message instead of falling back to the "no such file or directory" error